src/meta/src/controller/fragment.rs

```diff
@@ -1038,22 +1037,16 @@ impl CatalogController {
     pub async fn get_upstream_root_fragments(
         &self,
         upstream_job_ids: Vec<ObjectId>,
-        job_type: Option<PbTableJobType>,
     ) -> MetaResult<HashMap<ObjectId, PbFragment>> {
         let inner = self.inner.read().await;
 
         let mut fragments = Fragment::find()
             .filter(fragment::Column::JobId.is_in(upstream_job_ids))
             .all(&inner.db)
             .await?;
-        fragments.retain(|f| match job_type {
-            Some(PbTableJobType::SharedCdcSource) => {
-                f.fragment_type_mask & PbFragmentTypeFlag::Source as i32 != 0
-            }
-            // MV on MV, and other kinds of table job
-            None | Some(PbTableJobType::General) | Some(PbTableJobType::Unspecified) => {
-                f.fragment_type_mask & PbFragmentTypeFlag::Mview as i32 != 0
-            }
+        fragments.retain(|f| {
+            f.fragment_type_mask & PbFragmentTypeFlag::Source as i32 != 0
```

Comment:
A normal table also contains a `Source` fragment which should not filtered as the root fragment. That's why e2e fails I think.

Reply:
This makes me think what to do for MV on reuseable source..

Reply:
I came up with a solution: search for Source fragment only if there's no Mview.

Reply:
Actually we are looking for the bottom-most node in the graph. 🤔